src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs

```diff
@@ -407,142 +407,162 @@ internal virtual ImmutableArray<NamedTypeSymbol> GetAdditionalTopLevelTypes()
+                if (_lazyExportedTypes.Length > 0)
                 {
-                    var modules = sourceAssembly.Modules;
-                    for (int i = 1; i < modules.Length; i++) //NOTE: skipping modules[0]
-                    {
-                        GetExportedTypes(modules[i].GlobalNamespace, builder);
-                    }
+                    ReportExportedTypeNameCollisions(_lazyExportedTypes, diagnostics);
                 }
+            }
 
-                HashSet<NamedTypeSymbol> seenTopLevelForwardedTypes = new HashSet<NamedTypeSymbol>();
-                GetForwardedTypes(seenTopLevelForwardedTypes, sourceAssembly.GetSourceDecodedWellKnownAttributeData(), builder);
+            return _lazyExportedTypes;
+        }
 
-                if (!OutputKind.IsNetModule())
+        /// <summary>
+        /// Builds an array of public type symbols defined in netmodules included in the compilation
+        /// and type forwarders defined in this compilation or any included netmodule (in this order).
+        /// </summary>
+        private ImmutableArray<Cci.ExportedType> CalculateExportedTypes()
+        {
+            SourceAssemblySymbol sourceAssembly = SourceModule.ContainingSourceAssembly;
+            var builder = ArrayBuilder<Cci.ExportedType>.GetInstance();
+
+            if (!OutputKind.IsNetModule())
+            {
+                var modules = sourceAssembly.Modules;
+                for (int i = 1; i < modules.Length; i++) //NOTE: skipping modules[0]
                 {
-                    GetForwardedTypes(seenTopLevelForwardedTypes, sourceAssembly.GetNetModuleDecodedWellKnownAttributeData(), builder);
+                    GetExportedTypes(modules[i].GlobalNamespace, -1, builder);
                 }
+            }
 
-                Debug.Assert(_lazyExportedTypes.IsDefault);
+            var seenTopLevelForwardedTypes = new HashSet<NamedTypeSymbol>();
+            GetForwardedTypes(seenTopLevelForwardedTypes, sourceAssembly.GetSourceDecodedWellKnownAttributeData(), builder);
 
-                _lazyExportedTypes = builder.ToImmutableAndFree();
+            if (!OutputKind.IsNetModule())
+            {
+                GetForwardedTypes(seenTopLevelForwardedTypes, sourceAssembly.GetNetModuleDecodedWellKnownAttributeData(), builder);
+            }
 
-                if (_lazyExportedTypes.Length > 0)
+            return builder.ToImmutableAndFree();
+        }
+
+        private void ReportExportedTypeNameCollisions(ImmutableArray<Cci.ExportedType> exportedTypes, DiagnosticBag diagnostics)
+        {
+            var sourceAssembly = SourceModule.ContainingSourceAssembly;
+            var exportedNamesMap = new Dictionary<string, NamedTypeSymbol>(StringOrdinalComparer.Instance);
+
+            foreach (var exportedType in exportedTypes)
+            {
+                var type = (NamedTypeSymbol)exportedType.Type;
+
+                Debug.Assert(type.IsDefinition);
+
+                if (type.IsTopLevelType())
```

Comment:
Done.

---
In reply to: [67207148](https://github.com/dotnet/roslyn/pull/12004#discussion_r67207148) [](ancestors = 67207148)